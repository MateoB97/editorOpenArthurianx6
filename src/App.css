.Tarjeta {
  margin: 2vh 3vw;
  border: 1px solid black;
  border-radius: 12px;
  padding-block-end: 2vh;
  max-height: 900px;
  max-width: 650px;
  width: max-content;
  height: max-content;
}

.Tarjeta__header {
  background-color: gray;
  border-radius: 12px 12px 0 0;
  padding-inline: 2%;
  padding-block: 2%;
  color: white;
}

.Tarjeta__head {
  margin: 0 5%;
}

.Tarjeta__imagen {
  display: inline-block;
  max-height: 20%;
  max-width: 20%;
  height: 40px;
  width: 40px;
  margin: 5%;
  border: 0.5px solid black;
}

.Tarjeta__mobSelector {
  position: relative;
  left: 50%;
  width: 30%;
  max-width: 30%;
}

.Tarjeta__form {
  margin: 5%;
  max-height: 75%;
  height: 450px;
}

.Tarjeta__footer.-MobType {
  display: flex;
  align-items: center;
  justify-content: space-around;
}

.Formulario.-MobType {
  display: flex;
  flex-direction: column;
  flex-wrap: nowrap;
  justify-content: space-between;
  height: 65%;
  max-height: 65%;
  overflow-x: scroll;
}

.Formulario__campo.-MobType {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 15px;
  textarea {
    overflow-x: scroll;
    max-width: 50%;
    width: 50%;
    max-height: 20vh;
  }
}

.Items {
  height: min-content;
  height: 30%;
  max-height: 30%;
}

.Items__dropList.-MobType {
  border: 0.5px solid black;
  border-radius: 12px;
  padding-block-start: 10px;
  padding-block-end: 10px;
  padding-inline-start: 8px;
  max-height: 55%;
  height: 65px;
  overflow-x: scroll;
  list-style: none;
}

.Items__dropOption.-MobType {
  display: flex;
  justify-content: space-around;
  max-height: 15%;
  max-width: 90%;
  border: none;
  select {
    flex-basis: 20%;
    flex-grow: 2;
    flex-shrink: 1;
    max-width: 35%;
  }
  input {
    flex-basis: 10%;
    flex-grow: 1;
    flex-shrink: 2;
    max-height: min-content;
    max-width: 15%;
    width: 5%;
  }
}

.Lista__item.-DropList {
  display: flex;
  justify-content: space-between;
  margin: 2px 14px 4px 5px;
}

.Boton.-addLoot {
  /* position: relative;
  inset-inline: 40%; */
}

.Boton.-eliminar.-DropList {
  border-radius: 100%;
  border: 0.2 solid black;
  background-color: red;
  color: white;
  font-weight: 2rem;
  &:hover {
    cursor: pointer;
  }
}

.Boton.-exportar {
  /* position: relative;
  inset-inline: 40%;
  margin-top: 20px; */
}
