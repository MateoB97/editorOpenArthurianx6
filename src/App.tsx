import React, { RefObject } from "react";
import "./App.css";
import MobTypeData, { MobTemplate, LootItems } from "./data/MobTypes";
import Items from "./data/Items";

function App() {
  const MobID = React.useRef<HTMLSelectElement>(null);
  const typeNewItem = React.useRef<HTMLSelectElement>(null);
  const qtyNewItem = React.useRef<HTMLInputElement>(null);
  const lootItem = React.useRef<HTMLLIElement>(null);
  let MobObj: MobTemplate | undefined;
  const [MobGlobalData, setMobGlobal] =
    React.useState<Array<MobTemplate | undefined>>(MobTypeData);
  const [MobActual, setMobActual] = React.useState<MobTemplate | undefined>(
    MobObj,
  );

  function addValueInput(
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
  ) {
    let element = event.target;
  }

  function exportData(): void {
    const jsonString = `data:text/json;chatset=utf-8,${encodeURIComponent(
      JSON.stringify(MobGlobalData),
    )}`;
    const link = document.createElement("a");
    link.href = jsonString;
    link.download = "data.json";
    link.click();
  }

  function handleMobId(event: React.ChangeEvent<any>) {
    MobID.current!.value = event.target.value;
    setMobActual(
      MobGlobalData.find((mob) => mob?.id === MobID?.current?.value),
    );
  }

  function deleteLootMob(
    event: React.MouseEvent,
    id: string | undefined,
  ): void {
    event.stopPropagation();
    function filterItem(actual: MobTemplate | undefined) {
      let newItems: Array<LootItems>;
      if (actual !== undefined) {
        if (actual?.items) {
          newItems = actual?.items.filter((item) => id !== item.itemId);
          actual.items = newItems;
          return actual;
        }
      }
    }
    setMobActual(filterItem(MobActual));
  }

  // const deepMergeObj = (...objects: Array<any>) => {
  //   const CopiasObj = objects.map((obj) => JSON.parse(JSON.stringify(obj)));
  //   const Reduce = CopiasObj.reduce(
  //     (merged, current) => ({ ...current, ...merged }),
  //     {},
  //   );
  //   return Reduce;
  // };

  function addLootMob(): void {
    function pushItem(
      actual: MobTemplate | undefined,
    ): MobTemplate | undefined {
      const itemValue = typeNewItem.current?.value ?? "";
      const qtyValue = parseInt(qtyNewItem.current?.value ?? "", 10);
      let nuevoLoot: {
        itemId: string | undefined;
        quantity: number | undefined;
      } = {
        itemId: itemValue,
        quantity: qtyValue,
      };
      if (actual !== undefined) {
        if (!actual?.items) {
          actual!.items = [];
        }
        let lootFiltro = actual?.items!.filter(
          (item) => item.itemId !== nuevoLoot.itemId,
        );
        lootFiltro.push(nuevoLoot);
        actual.items = lootFiltro;
      }
      return actual;
    }
    setMobActual(pushItem(MobActual));
  }

  return (
    <article className="Tarjeta">
      <header className="Tarjeta__header">
        <h4>Monster Editor</h4>
      </header>
      <section className="Tarjeta__head">
        <figure className="Tarjeta__imagen">
          <img src="" alt="" id="moster[portrait]" />
        </figure>
        <select
          ref={MobID}
          onChange={handleMobId}
          className="Tarjeta__mobSelector"
          name="mob[id]"
          id=""
        >
          <option>Select</option>
          {MobGlobalData.map((mob, key) => {
            return (
              <option key={key} value={mob?.id}>
                {mob?.name}
              </option>
            );
          })}
        </select>
      </section>
      <section className="Tarjeta__form">
        <form className="Formulario -MobType" action="">
          <label className="Formulario__campo -MobType" htmlFor="">
            <span>Name</span>
            <input
              id="mob-name"
              name="mob[name]"
              onChange={addValueInput}
              value={MobActual?.name ?? ""}
              type="text"
            />
          </label>
          <label className="Formulario__campo -MobType" htmlFor="">
            <span>Appearance</span>
            <input
              id="mob-appearance"
              name="mob[appearance]"
              onChange={addValueInput}
              value={MobActual?.appearance ?? ""}
              type="text"
            />
          </label>
          <label className="Formulario__campo -MobType" htmlFor="">
            <span>Description</span>
            <textarea
              id="mob-description"
              onChange={addValueInput}
              value={MobActual?.description ?? ""}
              name="mob[description]"
              cols={22}
              rows={10}
            ></textarea>
          </label>
          <label className="Formulario__campo -MobType" htmlFor="">
            <span>Corpse</span>
            <input
              id="mob-corpse"
              onChange={addValueInput}
              name="mob[corpse]"
              type="text"
            />
          </label>
          <label className="Formulario__campo -MobType" htmlFor="">
            <span>Intent</span>
            <input
              id="mob-intent"
              onChange={addValueInput}
              name="mob[intent]"
              type="text"
            />
          </label>
          <label className="Formulario__campo -MobType" htmlFor="">
            <span>Weapon</span>
            <input
              id="mob-weapon"
              onChange={addValueInput}
              name="mob[weapon]"
              type="text"
            />
          </label>
          <label className="Formulario__campo -MobType" htmlFor="">
            <span>Alignment</span>
            <input
              id="mob-alignment"
              onChange={addValueInput}
              name="mob[alignment]"
              type="text"
            />
          </label>
          <label className="Formulario__campo -MobType" htmlFor="">
            <span>Damage</span>
            <input
              id="mob-damage"
              onChange={addValueInput}
              name="mob[damage]"
              type="number"
            />
          </label>
          <label className="Formulario__campo -MobType" htmlFor="">
            <span>Hit points</span>
            <input
              id="mob-hitPoints"
              onChange={addValueInput}
              name="mob[hitPoints]"
              type="number"
            />
          </label>
          <label className="Formulario__campo -MobType" htmlFor="">
            <span>Defense</span>
            <input
              id="mob-defense"
              onChange={addValueInput}
              name="mob[defense]"
              type="number"
            />
          </label>
          <label className="Formulario__campo -MobType" htmlFor="">
            <span>Speed</span>
            <input
              id="mob-speed"
              onChange={addValueInput}
              name="mob[speed]"
              type="number"
            />
          </label>
          <select
              ref={typeNewItem}
              name="mob[useEffect]"
              id="mob_useEffect"
            >
              <option>Select</option>
              
          </select>
        </form>
        <section className="Items">
          <ul className="Items__dropList -MobType" id="dropItemList">
            {MobActual?.items?.map((item, key) => {
              return (
                <li
                  className="Lista__item -DropList"
                  ref={lootItem}
                  key={item.itemId}
                >
                  <span>{`Drop ${item.quantity} ${item.itemId}`}</span>
                  <button
                    className="Boton -eliminar -DropList"
                    type="button"
                    onClick={(e) => deleteLootMob(e, item.itemId)}
                  >
                    <b>X</b>
                  </button>
                </li>
              );
            })}
          </ul>
          <fieldset className="Items__dropOption -MobType">
            <select
              ref={typeNewItem}
              name="mob[dropItem]"
              id="mob_dropItem"
            >
              <option>Select</option>
              {Items.map((item, key) => {
                return (
                  <option key={key} value={item.name}>
                    {item.name}
                  </option>
                );
              })}
            </select>
            <input
              ref={qtyNewItem}
              className=""
              type="number"
              name="mob[dropQty]"
              id="mob_dropItemQty"
            />
          </fieldset>
        </section>
      </section>
      <footer className="Tarjeta__footer -MobType">
        <button type="button" onClick={addLootMob} className="Boton -addLoot">
          Add Loot
        </button>
        <button className="Boton -exportar" type="button" onClick={exportData}>
          Exportar
        </button>
      </footer>
    </article>
  );
}
export default App;
