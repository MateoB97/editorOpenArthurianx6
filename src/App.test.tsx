import React from "react";
import { render, screen } from "@testing-library/react";
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import App from "./App";

describe('Snapshot Componente', () => {
  
  // Instantaneas

  it('renders correctly', () => {
    const tree = renderer
    .create(<App />)
    .toJSON();

    expect(tree).toMatchSnapshot();
  })

  test("render componente editor de mobs", async () => {
    render(<App />);
  });

  test("renderiza la lista de items", async () => {
    const { getByRole } = render(<App />);
    expect(getByRole('list')).toBeInTheDocument();
  });
  
  // Acciones Usuario
  
  test("puede add items" , async () => {
    const user = userEvent.setup()
    user.click(screen.getByRole('boton_addLoot'))
    // await 
    expect(screen.getByRole('boton_addLoot')).toBeInTheDocument();
  
  })

  test("puede delete items" , async () => {
    const user = userEvent.setup()
    // await user.click(screen.getByRole('boton_deleteLoot'))
  
    expect(screen.getByRole('boton_deleteLoot')).toBeInTheDocument();
  
  })
})
