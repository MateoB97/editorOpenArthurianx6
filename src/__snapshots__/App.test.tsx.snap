// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Componente renders correctly 1`] = `
<article
  className="Tarjeta"
>
  <header
    className="Tarjeta__header"
  >
    <h4>
      Monster Editor
    </h4>
  </header>
  <section
    className="Tarjeta__head"
  >
    <figure
      className="Tarjeta__imagen"
    >
      <img
        alt=""
        src=""
      />
    </figure>
    <select
      className="Tarjeta__mobSelector"
      id=""
      name="monster[id]"
      onChange={[Function]}
    >
      <option>
        Select
      </option>
      <option
        value="demon"
      >
        Gargoyle
      </option>
      <option
        value="avatar"
      >
        Avatar
      </option>
      <option
        value="man1"
      >
        Man
      </option>
      <option
        value="iolo"
      >
        Iolo
      </option>
      <option
        value="shamino"
      >
        Shamino
      </option>
      <option
        value="corvus"
      >
        Corvus
      </option>
      <option
        value="calista"
      >
        Calista
      </option>
      <option
        value="rat"
      >
        Rat
      </option>
      <option
        value="cow"
      >
        Cow
      </option>
      <option
        value="skeleton"
      >
        Skeleton
      </option>
      <option
        value="swordSoldier"
      >
        Soldier
      </option>
      <option
        value="xbowSoldier"
      >
        Soldier
      </option>
      <option
        value="asteroth"
      >
        Asteroth
      </option>
    </select>
  </section>
  <section
    className="Tarjeta__form"
  >
    <form
      action=""
      className="Formulario -MobType"
    >
      <label
        className="Formulario__campo -MobType"
        htmlFor=""
      >
        <span>
          Name
        </span>
        <input
          id="monster-name"
          name="monster[name]"
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
      <label
        className="Formulario__campo -MobType"
        htmlFor=""
      >
        <span>
          Description
        </span>
        <textarea
          cols={22}
          id="monster-description"
          name="monster[description]"
          onChange={[Function]}
          rows={10}
          value=""
        />
      </label>
      <label
        className="Formulario__campo -MobType"
        htmlFor=""
      >
        <span>
          Weight
        </span>
        <input
          id="monster-weight"
          name="monster[weight]"
          onChange={[Function]}
          type="number"
        />
      </label>
      <label
        className="Formulario__campo -MobType"
        htmlFor=""
      >
        <span>
          Hit points
        </span>
        <input
          id="monster-hitPoints"
          name="monster[hitPoints]"
          onChange={[Function]}
          type="number"
        />
      </label>
      <label
        className="Formulario__campo -MobType"
        htmlFor=""
      >
        <span>
          Attack
        </span>
        <input
          id="monster-attack"
          name="monster[attack]"
          onChange={[Function]}
          type="number"
        />
      </label>
      <ul
        className="Formulario__dropList -MobType"
        id="dropItemList"
      />
      <fieldset
        className="Formulario__dropOption -MobType"
      >
        <select
          id="monster_dropItem"
          name="monster[dropItem]"
        >
          <option>
            Select
          </option>
          <option
            value="Fish"
          >
            Fish
          </option>
          <option
            value="Lobster"
          >
            Lobster
          </option>
          <option
            value="Sword"
          >
            Sword
          </option>
          <option
            value="Mystic Sword"
          >
            Mystic Sword
          </option>
          <option
            value="Spear of Shamap"
          >
            Spear of Shamap
          </option>
          <option
            value="Chain Mail"
          >
            Chain Mail
          </option>
          <option
            value="Magic Axe"
          >
            Magic Axe
          </option>
          <option
            value="Crossbow"
          >
            Crossbow
          </option>
          <option
            value="Iron Bolt"
          >
            Iron Bolt
          </option>
          <option
            value="Boomerang"
          >
            Boomerang
          </option>
          <option
            value="Main Gauche"
          >
            Main Gauche
          </option>
          <option
            value="Demon Body"
          >
            Demon Body
          </option>
          <option
            value="Man Body"
          >
            Man Body
          </option>
          <option
            value="Rat Body"
          >
            Rat Body
          </option>
          <option
            value="Cow Body"
          >
            Cow Body
          </option>
          <option
            value="Bones"
          >
            Bones
          </option>
          <option
            value="Door"
          >
            Door
          </option>
          <option
            value="Door"
          >
            Door
          </option>
          <option
            value="Door"
          >
            Door
          </option>
          <option
            value="Door"
          >
            Door
          </option>
          <option
            value="Gold key"
          >
            Gold key
          </option>
          <option
            value="Journal of The Quest"
          >
            Journal of The Quest
          </option>
          <option
            value="Ginseng Potion"
          >
            Ginseng Potion
          </option>
          <option
            value="Blood Moss Potion"
          >
            Blood Moss Potion
          </option>
          <option
            value="Lute"
          >
            Lute
          </option>
          <option
            value="Harpsichord"
          >
            Harpsichord
          </option>
          <option
            value="Piano"
          >
            Piano
          </option>
          <option
            value="Torch"
          >
            Torch
          </option>
          <option
            value="Barrel"
          >
            Barrel
          </option>
          <option
            value="Backpack"
          >
            Backpack
          </option>
          <option
            value="Bag"
          >
            Bag
          </option>
          <option
            value="Rock"
          >
            Rock
          </option>
          <option
            value="Oak Tree"
          >
            Oak Tree
          </option>
          <option
            value="Carrot"
          >
            Carrot
          </option>
          <option
            value="Empty Bucket"
          >
            Empty Bucket
          </option>
          <option
            value="Bucket (Milk)"
          >
            Bucket (Milk)
          </option>
        </select>
        <input
          className=""
          id="monster_dropItemQty"
          name="monster[dropQty]"
          type="number"
        />
      </fieldset>
    </form>
  </section>
  <footer
    className="Tarjeta__footer"
  >
    <button
      className="Boton -addLoot"
      onClick={[Function]}
      role="boton_addLoot"
      type="button"
    >
      Add Loot
    </button>
  </footer>
  <button
    className="Boton -exportar"
    onClick={[Function]}
    type="button"
  >
    Exportar
  </button>
</article>
`;
